reputation/reputation.go

```diff
@@ -19,39 +25,165 @@ type ReputationModule struct {
+
+	lockScores sync.Mutex
+	rebuild    chan struct{}
+	scores     []MinerScore
+
 	ctx    context.Context
 	cancel context.CancelFunc
 }
 
-func New(ds datastore.TxnDatastore) *ReputationModule {
+type MinerScore struct {
+	Addr  string
+	Score int
+}
+
+func New(ds datastore.TxnDatastore, mi *miner.MinerIndex, si *slashing.SlashingIndex, ai *ask.AskIndex) *ReputationModule {
 	ctx, cancel := context.WithCancel(context.Background())
 	rm := &ReputationModule{
 		ds: ds,
+		mi: mi,
+		si: si,
+		ai: ai,
 
+		rebuild: make(chan struct{}, 1),
 		ctx:     ctx,
 		cancel:  cancel,
 		sources: source.NewStore(ds),
 	}
 
 	go rm.updateSources()
+	go rm.subscribeIndexes()
+	go rm.indexBuilder()
 
 	return rm
 }
 
+func (rm *ReputationModule) AddSource(id string, maddr ma.Multiaddr) error {
+	return rm.sources.Add(source.Source{Id: id, Maddr: maddr})
+}
+
+func (rm *ReputationModule) GetTopMiners(n int) ([]MinerScore, error) {
+	mr := make([]MinerScore, 0, n)
+	rm.lockScores.Lock()
+	for i := 0; i < n; i++ {
+		mr = append(mr, rm.scores[i])
+	}
+	rm.lockScores.Unlock()
+	return mr, nil
+}
+
 func (rm *ReputationModule) Close() error {
 	rm.cancel()
 	return nil
 }
 
-func (rm *ReputationModule) AddSource(id string, maddr ma.Multiaddr) error {
-	return rm.sources.Add(source.Source{Id: id, Maddr: maddr})
+func (rm *ReputationModule) subscribeIndexes() {
+	subMi := rm.mi.Listen()
+	subSi := rm.si.Listen()
+	subAi := rm.ai.Listen()
+
+	for {
+		rm.lockIndex.Lock()
+		select {
+		case <-rm.ctx.Done():
+			log.Info("terminating background index update")
+			rm.lockIndex.Unlock()
+			return
+		case <-subMi:
+			rm.mIndex = rm.mi.Get()
+		case <-subSi:
+			rm.sIndex = rm.si.Get()
+		case <-subAi:
+			rm.aIndex = rm.ai.Get()
+		}
+		rm.lockIndex.Unlock()
+		select {
+		case rm.rebuild <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (rm *ReputationModule) indexBuilder() {
+	for range rm.rebuild {
+		log.Debug("rebuilding index")
+		start := time.Now()
+
+		sources, err := rm.sources.GetAll()
+		if err != nil {
+			log.Errorf("error when getting sources: %s", err)
+			return
+		}
+		rm.lockIndex.Lock()
+		minerIndex := rm.mIndex
+		slashIndex := rm.sIndex
+		askIndex := rm.aIndex
+		rm.lockIndex.Unlock()
+
+		scores := make([]MinerScore, 0, len(minerIndex.Miners))
+		for addr := range minerIndex.Miners {
+			score := calculateScore(addr, minerIndex, slashIndex, askIndex, sources)
+			scores = append(scores, score)
+		}
+		sort.Slice(scores, func(i, j int) bool {
+			return scores[i].Score > scores[j].Score
+		})
+
+		rm.lockScores.Lock()
+		rm.scores = scores
+		rm.lockScores.Unlock()
+
+		log.Debug("index rebuilt int %dms", time.Since(start).Milliseconds())
+	}
+}
+
+func calculateScore(addr string, mi miner.Index, si slashing.Index, ai ask.Index, ss []source.Source) MinerScore {
+	minfo := mi.Miners[addr]
+	powerScore := minfo.Power.Relative
+
+	var slashScore float64
+	if si, ok := si.Miners[addr]; ok {
+		slashScore = 1 / math.Pow(2, float64(len(si.History)))
```

Comment:
So, slash score non-linearly decreases with count... harsh! :) I have zero intuitions here.

Reply:
Yep, it feels to me that shouldn't be linear.
I think a better thing would be to also consider how many deals has on-chain.
If a miner has 3 deals, or 10000 deals... feels like it should have to be normalized or something similar. 
So maybe first normalize, then being exponentially penalized :P

Reply:
Ah yeah, makes sense.